articles/reproducibility/index.md

```diff
@@ -0,0 +1,118 @@
+
+**Skill versus randomness**. Consider a context where creators of machine learning models may claim to have improved the overall performance by tweaking a parameter. The performance may improve, but it might be impossible to tell whether the improvement is due to scientific skill or some randomness source. The only way to validate the improvement is to replicate the process.
+
+### Challenges to Attaining Reproducibility
+
+There exist many factors that hinder the efforts to attain reproducibility. Let's go over a few of them.
+
+#### Data
+
+Data is a key input to machine learning algorithms. These models are highly dependent on the data that is used to train them. Therefore, data that is fed into the model has not only a significant impact on performance but also reproducibility. A model will only produce the same results as another if the same data and flows are used to train it.
+
+Researchers may use a training set that programmers/data scientists attempting to implement the model may not have access to in a research context. Furthermore, some research databases rewrite or regularly make updates to data. This means that the training set for a given algorithm may not be available after a given period.
+
+#### Concept Drift
+
+One may struggle to deploy a model more than once. A key reason for this is concept drift. [Concept drift](/engineering-education/correcting-data-shift/) refers to the shift in the relationships between input and output variables in a given problem. Data power our models. But if the distribution of this data keeps changing, it proves to be a challenge for reproducibility. For example, consumer habits may be reflected by data that we seek to use to train our model. However, consumer habits always change with time. If someone wanted to reproduce our model and get similar results using up-to-date data, it would be a tall order.
+
+#### Algorithm
+
+Deterministic algorithms produce the same output given the same input, even on different iterations. However, there exist many algorithms that display non-determinism. Reproducibility in such algorithms is a difficult task since we expect randomness. Non-deterministic algorithms may be fed with the same input yet produce different results upon different iterations. The randomness in results may result from the algorithm using various paths to arrive at contrasting outcomes. A huge factor that influences the non-deterministic nature of an algorithm is the complexity of the model. The routes and processes used to achieve output may be non-reproducible, therefore ensuring an element of randomness. It is impossible to reproduce that which is random.
```

Comment:
How do we deal with non-deterministic algorithms? How do we ensure reproducibility in this case?